IdentityCore/src/cache/MSIDMacKeychainTokenCache.m

```diff
@@ -0,0 +1,660 @@
+// with other applications from the same vendor, the app will need to specify the
+// shared group here. If set to 'nil' the main bundle's identifier will be used instead.
+//
+- (nullable instancetype)initWithGroup:(nullable NSString *)keychainGroup
+{
+    MSID_TRACE;
+
+    self = [super init];
+    if (self)
+    {
+        if (!keychainGroup)
+        {
+            keychainGroup = [[NSBundle mainBundle] bundleIdentifier];
+        }
+
+        if (MSIDKeychainUtil.teamId)
+        {
+            // Add team prefix to keychain group if it is missed.
+            if (![keychainGroup hasPrefix:MSIDKeychainUtil.teamId])
+            {
+                keychainGroup = [MSIDKeychainUtil accessGroup:keychainGroup];
+            }
+        }
+
+        self.keychainGroup = keychainGroup;
+        if (!self.keychainGroup)
+        {
+            return nil;
+        }
+
+        self.defaultAccountQuery = @{
+                                     // All account items are saved as generic passwords.
+                                     (id)kSecClass: (id)kSecClassGenericPassword,
+
+                                     // Add the access group as it's own field for query filtering.
+                                     // Since the attribute is an NSNumber, hash the string.
+                                     (id)kSecAttrCreator: [NSNumber numberWithUnsignedInt:(uint32_t)self.keychainGroup.hash],
+
+                                     // Add a marker for all cache items in the keychain for additional query filtering.
+                                     (id)kSecAttrLabel: s_defaultKeychainLabel
+
+                                     };
+
+        MSID_LOG_INFO(nil, @"Init MSIDMacKeychainTokenCache with keychainGroup: %@", [self keychainGroupLoggingName]);
+        MSID_LOG_INFO_PII(nil, @"Init MSIDMacKeychainTokenCache with keychainGroup: %@", self.keychainGroup);
+    }
+
+    return self;
+}
+
+#pragma mark - Accounts
+
+// Write an account to the macOS keychain cache.
+//
+// Errors:
+// * MSIDErrorDomain/MSIDErrorInternal: invalid key or json deserialization error
+// * MSIDKeychainErrorDomain/OSStatus: Apple status codes from SecItemUpdate()/SecItemAdd()
+//
+- (BOOL)saveAccount:(MSIDAccountCacheItem *)account
+                key:(MSIDCacheKey *)key
+         serializer:(id<MSIDAccountItemSerializer>)serializer
+            context:(id<MSIDRequestContext>)context
+              error:(NSError **)error
+{
+    MSID_TRACE;
+    MSID_LOG_VERBOSE(context, @"Set keychain item, key info (account: %@ service: %@, keychainGroup: %@)", _PII_NULLIFY(key.account), _PII_NULLIFY(key.service), [self keychainGroupLoggingName]);
+    MSID_LOG_VERBOSE_PII(context, @"Set keychain item, key info (account: %@ service: %@, keychainGroup: %@)", key.account, key.service, self.keychainGroup);
+
+    if (!key.service.length)
+    {
+        [self createError:@"Set keychain item with invalid key (service is nil)."
```

Comment:
nit: is this extra helper necessary? Doesn't seem to do much.

Reply:
It isn't required. It just reduces some repetitive error-handling code.